### PR TITLE
feat: add cairo compilation to ci

### LINKED ISSUE #36: dev: Add cairo compilation to github CI

### PR DESCRIPTION
**Pull request by ftupas**
Fix #36

- Add cairo compilation to CI
- Add format checks
- Update protostar and cairo-lang versions

### CODE REVIEW

**Comment by maxgillett** [comment]
Hi @ftupas, thanks for working on this! A couple notes below, which others should definitely chime in on.

I tried compiling the Cairo code using the new Makefile and got a new error: `src/utxo_set/utxo_set.cairo:1:42: Redefinition of '__main__.alloc'`. It looks like the compiler is now trying to merge all into the same namespace and therefore complains any time that duplicate imports exist across the set of source code files. I don't think that we should remove these imports, and it makes me question why we were compiling all of the source files into a single JSON file in the first place. If the goal is to run the chain prover, then we need only pass a single Cairo source code file with a "main" entrypoint to the compiler, which will compile imported dependencies as needed. If there exists more than one entrypoint, the compiler should be called separately for each of those files containing `main` functions, as they are separate programs. If the goal is to compile all of the various subprograms/routines that might be of interest in their own right, or to catch bugs in currently unimported dependencies, then I think we should revert to the old compilation command but have each program outputted into a separate file in the `build` directory.

**Comment by ftupas** [comment]
Hi @maxgillett ! Thanks for the review, wishing you well for the holidays. My bad I thought the old command was trying to merge everything as the output was a single JSON file but now that I see each file has its own `main` entrypoint then it makes more sense to compile them separately. I'll revert to the old compilation and do it for each program so there will be separate files in the `build` directory 👍

File: Makefile [hunk 1]
```diff
@@ -4,13 +4,19 @@ BIN_DIR = ./bin
...
 RUST_TEST_LIB = $(BIN_DIR)/libzerosync_tests.dylib
 
-CAIRO_PROGRAM:
-	find src -type f \( -iname "*.cairo" -and -not -iname "test_*.cairo" \) \
-		-exec cairo-compile {} --cairo_path src > build/zerosync_compiled.json \;
+CAIRO_PROGRAMS:
+	@echo "Compiling Cairo files..."
+	@find src -type d | sed 's/src/build/g' | xargs mkdir -p
+	@for file in ${CAIRO_FILES}; do \
+		target=$$(echo "$${file%.cairo}_compiled.json" | sed 's/src/build/g'); \
```
Comment (ftupas) [inline Makefile:16]: This creates the sub-directories in `build` folder to mimic `src` structure and add `_compiled` suffix to compiled program i.e. `build/stark_verifier/composer_compiled.json`

File: Makefile [hunk 2]
```diff
@@ -4,13 +4,19 @@ BIN_DIR = ./bin
...
+		echo $$target...; \
+		cairo-compile $$file --cairo_path src --output $$target || exit 1; \
```
Comment (ftupas) [inline Makefile:18]: `|| exit 1` forces exit code 1 for compilation issues

**Comment by maxgillett** [comment]
Thanks @ftupas!

**Pull request merged**